ratis-server/src/main/java/org/apache/ratis/server/storage/FileChunkReader.java

```diff
@@ -50,11 +50,11 @@ public class FileChunkReader implements Closeable {
    * @param directory the directory where the file is stored.
    * @throws IOException if it failed to open the file.
    */
-  public FileChunkReader(FileInfo info, RaftStorageDirectory directory) throws IOException {
+  public FileChunkReader(FileInfo info, File directory) throws IOException {
     this.info = info;
     this.relativePath = Optional.of(info.getPath())
         .filter(Path::isAbsolute)
-        .map(p -> directory.getRoot().toPath().relativize(p))
+        .map(p -> directory.toPath().relativize(p))
```

Comment:
Thanks for reminding me of incompatibility issues and providing the patch!
I have a small question. Is it possible that we also let StateMachine provide `RootDir`? In this way, we can still have a unified handling logic in receiver end `SnapshotManager`, while we don't have to add `sm` prefix to the sender side. However, this may introduce a new restriction on StateMachine: the `snapshotDir` must be a children of `rootDir`. What do you think?

Reply:
In InstallSnapshotRequests.java, we can use `rootDir` to eliminate `sm`
```java
final File rootDir = server.getStateMachine().getStateMachineStorage().getRootDir();
final Function<Path, Path> relativize;
final Path dir;
if (rootDir != null) {
  dir = rootDir.toPath();
} else {
  dir = server.getRaftStorage().getStorageDir().getRoot().toPath();
}
relativize = dir::relativize;
```

In SnapshotManager.java
```java
final File snapshotDir = Optional.ofNullable(smStorage.getSnapshotDir()).orElseGet(dir::getStateMachineDir);
final File tmp =  Optional.ofNullable(smStorage.getTmpDir()).orElseGet(dir::getTmpDir);
final File rootDir = Optional.ofNullable(smStorage.getRootDir()).orElseGet(dir::getRoot);

this.getRelativePath = c -> snapshotDir.toPath().relativize(new File(rootDir, c.getFilename()).toPath()).toString();
```

Reply:
@SzyWilliam , we cannot change the path format in the proto.  Otherwise, it won't work if an old server/client is talking to a new client/server.  Therefore, we probably cannot remove the `sm/` prefix from the filename in the proto.

Reply:
BTW, compatibility is important for supporting rolling upgrade.

Reply:
@szetszwo Thanks for the explanation! I see, in this way, Ratis can automatically handle the compatibility issue between the old StateMachine using the default path and the new StateMachine using the customized path and requires no changes on StateMachine. Bravo!